controllers/clusters/on_premises.go

```diff
@@ -197,198 +228,220 @@ func reconcileSSHGatewayResources(ctx context.Context, b *onPremisesBootstrap) e
-			Namespace: b.K8sObject.GetNamespace(),
-			Name:      nodeName,
-		}, nodeVM)
-		if client.IgnoreNotFound(err) != nil {
-			return err
-		}
-		if k8serrors.IsNotFound(err) {
-			nodeVM, err = newVM(
-				ctx,
-				b,
-				nodeName,
-				node.ID,
-				node.Rack,
-				nodeOSDV.Name,
-				nodeCPU,
-				nodeMemory,
-				nodeDataDV.Name,
-			)
+	return &vmList.Items[0], nil
+}
+
+func reconcileDV(ctx context.Context, b *onPremisesBootstrap, nodeID, diskName, diskRole string) (*cdiv1beta1.DataVolume, error) {
+	dvList := &cdiv1beta1.DataVolumeList{}
+	err := b.K8sClient.List(ctx, dvList, &client.ListOptions{
+		LabelSelector: labels.SelectorFromSet(map[string]string{
+			models.NodeIDLabel: nodeID,
+			models.DVRoleLabel: diskRole,
+		}),
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	dv := &cdiv1beta1.DataVolume{}
+	if len(dvList.Items) == 0 {
+		dvSize := resource.Quantity{}
+		switch diskRole {
+		case models.OSDVLabel:
+			dvSize, err = resource.ParseQuantity(b.OnPremisesSpec.OSDiskSize)
 			if err != nil {
-				return err
+				return nil, err
 			}
-			err = b.K8sClient.Create(ctx, nodeVM)
+		case models.StorageDVLabel:
+			dvSize, err = resource.ParseQuantity(b.OnPremisesSpec.DataDiskSize)
 			if err != nil {
-				return err
+				return nil, err
 			}
 		}
 
-		if !b.PrivateNetworkCluster {
-			nodeExposeName := fmt.Sprintf("%s-%s", models.NodeSvcPrefix, nodeName)
-			nodeExposeService := &k8scorev1.Service{}
-			err = b.K8sClient.Get(ctx, types.NamespacedName{
-				Namespace: b.K8sObject.GetNamespace(),
-				Name:      nodeExposeName,
-			}, nodeExposeService)
-			if client.IgnoreNotFound(err) != nil {
-				return err
-			}
-			if k8serrors.IsNotFound(err) {
-				nodeExposeService = newExposeService(
-					b,
-					nodeExposeName,
-					nodeName,
-					node.ID,
-				)
-				err = b.K8sClient.Create(ctx, nodeExposeService)
-				if err != nil {
-					return err
-				}
-			}
+		dv, err = createDV(
+			ctx,
+			b,
+			diskName,
+			nodeID,
+			diskRole,
+			dvSize,
+		)
+		if err != nil {
+			return nil, err
 		}
 
-		headlessServiceName := fmt.Sprintf("%s-%s", models.KubevirtSubdomain, strings.ToLower(b.K8sObject.GetName()))
-		headlessSVC := &k8scorev1.Service{}
-		err = b.K8sClient.Get(ctx, types.NamespacedName{
-			Namespace: b.K8sObject.GetNamespace(),
-			Name:      headlessServiceName,
-		}, headlessSVC)
+		return dv, nil
+	}
 
-		if client.IgnoreNotFound(err) != nil {
-			return err
-		}
-		if k8serrors.IsNotFound(err) {
-			headlessSVC = newHeadlessService(
+	return &dvList.Items[0], nil
+}
+
+func reconcileService(ctx context.Context, b *onPremisesBootstrap, nodeID, nodeName, role string) (*k8scorev1.Service, error) {
+	svcList := &k8scorev1.ServiceList{}
+	svcLabels := map[string]string{
+		models.ServiceRoleLabel: role,
+		models.ClusterIDLabel:   b.ClusterStatus.ID,
+	}
+
+	var svcName string
+	if role == models.HeadlessServiceRoleLabel {
+		svcName = fmt.Sprintf("%s-%s", role, strings.ToLower(b.K8sObject.GetName()))
```

Comment:
Why do you need this? `strings.ToLower(b.K8sObject.GetName()))` it should also work with identical name

Reply:
You are right, I can just use `GetName()`